### PR TITLE
Fix not processing enums in cpp-pistache-server

### LINKED ISSUE #1317: [cpp-pistache-server] Enumerations not generated
##### Description

Currently the enumeration definition below generates a class with no members.  The attached file archive.zip contains a complete definition (with several enumerations) that demonstrate this problem.

[Archive.zip](https://github.com/OpenAPITools/openapi-generator/files/2515830/Archive.zip)

##### openapi-generator version

v 3.3.0

##### OpenAPI declaration file content or url

```yaml
    SupportedGADShapes:
      anyOf:
        - type: string
          enum:
            - POINT
            - POINT_UNCERTAINTY_CIRCLE
            - POINT_UNCERTAINTY_ELLIPSE
            - POLYGON
            - POINT_ALTITUDE
            - POINT_ALTITUDE_UNCERTAINTY
            - ELLIPSOID_ARC
        - type: string
```

##### Command line used for generation

java -jar /home/brian/code/git/openapi-generator/modules/openapi-generator-cli/target/openapi-generator-cli.jar generate -i updated/TS29122_MonitoringEvent.yaml -g cpp-pistache-server --config __config__ -o ./code/MonitoringEvent/server/cpp-pistache-server

##### Steps to reproduce

1. Create a blank working directory
2. Unzip Archive.zip to the newly created directory.
3. Generate the source using the provided command making path adjustments for your system.
4. The class SupportedGADShapes in ./code/MonitoringEvent/server/cpp-pistache-server/model/SupportedGADShapes.h has no members.

##### Suggest a fix/enhancement

The attached file SupportedGADShapes.zip contains the .h and .cpp definitions for the enumeration. 
 Note that an enumeration named "INVALID" has been added in the header and is used when no default is specified.

[SupportedGADShapes.cpp-pistache-server.zip](https://github.com/OpenAPITools/openapi-generator/files/2515849/SupportedGADShapes.cpp-pistache-server.zip)

```
...
    /////////////////////////////////////////////
    /// SupportedGADShapes members
    enum eSupportedGADShapes
    {
       INVALID,
       POINT,
       POINT_UNCERTAINTY_CIRCLE,
       POINT_UNCERTAINTY_ELLIPSE,
       POLYGON,
       POINT_ALTITUDE,
       POINT_ALTITUDE_UNCERTAINTY,
       ELLIPSOID_ARC
    };

    eSupportedGADShapes getValue() const;
    void setValue(eSupportedGADShapes const value);

protected:
    eSupportedGADShapes m_value;
...
```

```
...
nlohmann::json SupportedGADShapes::toJson() const
{
    nlohmann::json val;
    
    switch (m_value)
    {
       case eSupportedGADShapes::POINT:                        val = "POINT";  break;
       case eSupportedGADShapes::POINT_UNCERTAINTY_CIRCLE:     val = "POINT_UNCERTAINTY_CIRCLE";  break;
       case eSupportedGADShapes::POINT_UNCERTAINTY_ELLIPSE:    val = "POINT_UNCERTAINTY_ELLIPSE";  break;
       case eSupportedGADShapes::POLYGON:                      val = "POLYGON";  break;
       case eSupportedGADShapes::POINT_ALTITUDE:               val = "POINT_ALTITUDE";  break;
       case eSupportedGADShapes::POINT_ALTITUDE_UNCERTAINTY:   val = "POINT_ALTITUDE_UNCERTAINTY";  break;
       case eSupportedGADShapes::ELLIPSOID_ARC:                val = "ELLIPSOID_ARC";  break;
       default:                                                val = "INVALID";  break;
    }

    return val;
}

void SupportedGADShapes::fromJson(nlohmann::json& val)
{
    auto s = val.get<std::string>();
    
    m_value =
          (s == "POINT")                        ? eSupportedGADShapes::POINT :
          (s == "POINT_UNCERTAINTY_CIRCLE")     ? eSupportedGADShapes::POINT_UNCERTAINTY_CIRCLE :
          (s == "POINT_UNCERTAINTY_ELLIPSE")    ? eSupportedGADShapes::POINT_UNCERTAINTY_ELLIPSE :
          (s == "POLYGON")                      ? eSupportedGADShapes::POLYGON :
          (s == "POINT_ALTITUDE")               ? eSupportedGADShapes::POINT_ALTITUDE :
          (s == "POINT_ALTITUDE_UNCERTAINTY")   ? eSupportedGADShapes::POINT_ALTITUDE_UNCERTAINTY :
          (s == "ELLIPSOID_ARC")                ? eSupportedGADShapes::ELLIPSOID_ARC :
                eSupportedGADShapes::INVALID;
}

SupportedGADShapes::eSupportedGADShapes SupportedGADShapes::getValue() const
{
   return m_value;
}

void SupportedGADShapes::setValue(SupportedGADShapes::eSupportedGADShapes const value)
{
   m_value = value;
}
...
```

### PR DESCRIPTION
**Pull request by mfyuce**
Defining a reusable enum as a component schema results in a class that does not include enum values. Following changes are made:

- override 'toEnumValue' in 'AbstractCppCodegen'
- modify model templates for the 'cpp-pistache-server' project
 

Fixes #1317

### PR checklist
 
- [ x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [ x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x ] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [ x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `5.1.x`, `6.0.x`
- [ x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

@ravinikam (2017/07) @stkrwork (2017/07) @etherealjoy (2018/02) @martindelille (2018/03) @muttleyxd (2019/08)

### CODE REVIEW

**Comment by wing328** [comment]
@mfyuce thanks for the PR. Is it correct to say that you've tested these changes locally and it works for your use cases?

File: modules/openapi-generator/src/main/resources/cpp-pistache-server/model-source.mustache [hunk 1]
```diff
@@ -32,6 +34,20 @@ void to_json(nlohmann::json& j, const {{classname}}& o)
...
+        {{#-first}}
+        case {{classname}}::e{{classname}}::INVALID_VALUE_OPENAPI_GENERATED:
+            j = "INVALID_VALUE_OPENAPI_GENERATED";
+            break;
+        {{/-first}}
+        case {{classname}}::e{{classname}}::{{name}}:
+            j = "{{value}}";
+            break;
+        {{/enumVars}}
+}{{/allowableValues}}{{/isEnum}}
```
Comment (etherealjoy) [inline modules/openapi-generator/src/main/resources/cpp-pistache-server/model-source.mustache:50]: Indent here.

**Review by etherealjoy — APPROVED** [review]

**Comment by etherealjoy** [comment]
I tried it and it seems OK. 
There is a limitation still, in case an enum is part of the class it/nested inside it, is not generated. 
But this can be handled later in a future PR.

File: modules/openapi-generator/src/main/resources/cpp-pistache-server/model-source.mustache (hunk 1 above)
Comment (mfyuce) [reply modules/openapi-generator/src/main/resources/cpp-pistache-server/model-source.mustache]: done and committed to the PR

**Comment by mfyuce** [comment]
Hi @wing328,

>Is it correct to say that you've tested these changes locally and it works for your use cases?

Yes, I did the testes for our use cases and it is usable for us.

But, after server generation, we needed to do a bit manuel modification and then we were ok (delete all "**EnumContainer**"s, rename all "**EnumContainer_anyOf**"s to "**EnumContainer**").

May be @etherealjoy is telling about the same situation. 

> There is a limitation still, in case an enum is part of the class it/nested inside it, is not generated.
> But this can be handled later in a future PR.

I checked qt server generation output and saw that the same situation is there, so, I thought it is a feature or sth...

**Comment by wing328** [comment]
CI failure not related to this change.

**Comment by wing328** [comment]
@mfyuce thanks for the PR, which has been included in the v5.1.0 release: https://twitter.com/oas_generator/status/1373636654024380423

**Comment by mfyuce** [comment]
@wing328 thanks to you and to the team for the great tool and the work!

**Pull request merged**